Sources/Profiler.swift

```diff
@@ -0,0 +1,326 @@
+        guard started.pending else { return self }
+        return PendingResult(created: created, identity: identity, attached: attached, started: .ready(now - created), cancelled: cancelled, finished: finished, children: children)
+    }
+
+    func cancel(now: TimeInterval = CFAbsoluteTimeGetCurrent() as TimeInterval) -> PendingResult {
+        guard cancelled.pending else { return self }
+        return PendingResult(created: created, identity: identity, attached: attached, started: started, cancelled: .ready(now - created), finished: finished, children: children)
+    }
+
+    func finish(now: TimeInterval = CFAbsoluteTimeGetCurrent() as TimeInterval) -> PendingResult {
+        guard finished.pending else { return self }
+        return PendingResult(created: created, identity: identity, attached: attached, started: started, cancelled: cancelled, finished: .ready(now - created), children: children)
+    }
+
+    func addChild(child: ProfileResult) -> PendingResult {
+        var newChildren = children
+        newChildren.append(child)
+        return PendingResult(created: created, identity: identity, attached: attached, started: started, cancelled: cancelled, finished: finished, children: newChildren)
+    }
+}
+
+// MARK: ProcedureProfiler
+public final class ProcedureProfiler: Identifiable, Equatable {
+
+    public let identifier = UUID()
+
+    enum Reporter {
+        case Parent(ProcedureProfiler)
+        case Reporters([ProcedureProfilerReporter])
+    }
+
+
+    let queue = DispatchQueue(label: "me.danthorpe.ProcedureKit.Profiler")
+    let reporter: Reporter
+
+    var result = PendingResult(created: CFAbsoluteTimeGetCurrent() as TimeInterval, identity: .pending, attached: .pending, started: .pending, cancelled: .pending, finished: .pending, children: [])
+    var children: [Procedure.Identity] = []
+    var finishedOrCancelled = false
+
+    var pending: Bool {
+        return result.pending || (children.count > 0)
+    }
+
+    public convenience init(reporters: [ProcedureProfilerReporter]) {
+        self.init(reporter: .Reporters(reporters))
+    }
+
+    convenience init(parent: ProcedureProfiler) {
+        self.init(reporter: .Parent(parent))
+    }
+
+    init(reporter: Reporter) {
+        self.reporter = reporter
+    }
+
+    func addMetricNow(now: TimeInterval = CFAbsoluteTimeGetCurrent() as TimeInterval, forEvent event: ProcedureEvent) {
+        queue.sync { [unowned self] in
+            switch event {
+            case .attached:
+                self.result = self.result.attach(now: now)
+            case .started:
+                self.result = self.result.start(now: now)
+            case .cancelled:
+                self.result = self.result.cancel(now: now)
+                self.finishedOrCancelled = true
+            case .finished:
+                self.result = self.result.finish(now: now)
+                self.finishedOrCancelled = true
```

Comment:
I'm thinking that because this is a `.sync` method, that we don't need to capture `self` and can just clean this up to be:

```swift
queue.sync {
    switch event {
        case .attached:
            result = result.attach(now: now)
        // etc
```